python/ray/serve/controller.py

```diff
@@ -408,25 +409,87 @@ def deploy_app(
             self.deployment_timestamp = time.time()
 
         config_dict = config.dict(exclude_unset=True)
+
+        # Compare new config options with old ones and set versions of new deployments
+        config_checkpoint = self.kv_store.get(CONFIG_CHECKPOINT_KEY)
+        if config_checkpoint is not None:
+            _, last_config_dict, last_version_dict = pickle.loads(config_checkpoint)
+            version_dict = self.version_config(
+                config_dict, last_config_dict, last_version_dict
+            )
+        else:
+            version_dict = self.version_config(config_dict, {}, {})
+
         self.kv_store.put(
             CONFIG_CHECKPOINT_KEY,
-            pickle.dumps((self.deployment_timestamp, config_dict)),
+            pickle.dumps((self.deployment_timestamp, config_dict, version_dict)),
         )
 
+        deployment_override_options = config_dict.get("deployments", [])
+
         if self.config_deployment_request_ref is not None:
             ray.cancel(self.config_deployment_request_ref)
             logger.info(
                 "Received new config deployment request. Cancelling "
                 "previous request."
             )
 
-        deployment_override_options = config.dict(
-            by_alias=True, exclude_unset=True
-        ).get("deployments", [])
-
         self.config_deployment_request_ref = run_graph.options(
             runtime_env=config.runtime_env
-        ).remote(config.import_path, config.runtime_env, deployment_override_options)
+        ).remote(
+            config.import_path,
+            config.runtime_env,
+            deployment_override_options,
+            version_dict,
+        )
+
+    def version_config(
+        self, new_config: Dict, old_config: Dict, version_dict: Dict
+    ) -> Dict[str, str]:
+        """
+        When a deployment's options change, its version should generally change,
+        so old replicas are torn down. The only options which can be updated
+        without tearing down replicas (i.e. changing the version) are:
+
+        * num_replicas
+        * user_config
+        * autoscaling_config
+
+        This function determines whether each deployment's version
+        should be changed based on the updated options.
+        """
```

Comment:
This method needs a little bit more work before merging. Can we do the following:
- This looks like a pure function instead of method (not using `self`), consider moving this out the class, and into a function in the same file (lead with _ to indicate private function
- Start the docstring describing what does this function do. Essentially the last sentence.
- What exactly is the return type here? Is it a new copy of the version_dict? What's the Dict's key and value representing?

Reply:
Thanks, addressed!